pkcs11uri_test.go

```diff
@@ -78,27 +78,73 @@ func verifyPIN(t *testing.T, uri *Pkcs11URI, expectedpin string) {
 	}
 }
 
+func testConstruct(t *testing.T, uri *Pkcs11URI, expecteduri string) {
+	orig_expecteduri := expecteduri
+
+	verifyURI(t, uri, expecteduri)
+
+	expectedpin := "1234"
+	expecteduri = orig_expecteduri + fmt.Sprintf("?pin-value=%s", expectedpin)
+
+	err := uri.AddQueryAttribute("pin-value", expectedpin)
+	if err != nil {
+		t.Fatalf("Could not add query attribute: %s", err)
+	}
+
+	verifyURI(t, uri, expecteduri)
+	verifyPIN(t, uri, expectedpin)
+
+	expecteduri = orig_expecteduri + "?pin-value=the%20pin"
+	binarypin := []byte("the pin")
+
+	uri.RemoveQueryAttribute("pin-value")
+	err = uri.AddQueryAttributeUnencoded("pin-value", binarypin)
+	if err != nil {
+		t.Fatalf("Could not add unencoded query attribute: %s", err)
+	}
+
+	verifyURI(t, uri, expecteduri)
+	verifyPIN(t, uri, string(binarypin))
+}
+
 func TestConstruct1(t *testing.T) {
 	uri := New()
-	expecteduri := "pkcs11:id=%66%6F%6F"
 
 	err := uri.AddPathAttribute("id", "%66oo")
 	if err != nil {
 		t.Fatalf("Could not add path attribute: %s", err)
 	}
+	testConstruct(t, uri, "pkcs11:id=%66%6F%6F")
+}
 
-	verifyURI(t, uri, expecteduri)
+func TestConstruct2(t *testing.T) {
+	uri := New()
 
-	expectedpin := "1234"
-	expecteduri += fmt.Sprintf("?pin-value=%s", expectedpin)
+	err := uri.AddPathAttributeUnencoded("id", []byte{0x66,0x6f,0x6f})
+	if err != nil {
+		t.Fatalf("Could not add path attribute: %s", err)
+	}
+	testConstruct(t, uri, "pkcs11:id=%66%6F%6F")
+}
 
-	err = uri.AddQueryAttribute("pin-value", expectedpin)
+func TestConstruct3(t *testing.T) {
+	uri := New()
+
+	err := uri.AddPathAttribute("slot-id", "12345")
 	if err != nil {
-		t.Fatalf("Could not add query attribute: %s", err)
+		t.Fatalf("Could not add path attribute: %s", err)
 	}
+	testConstruct(t, uri, "pkcs11:slot-id=12345")
+}
 
-	verifyURI(t, uri, expecteduri)
-	verifyPIN(t, uri, expectedpin)
+func TestConstruct4(t *testing.T) {
+	uri := New()
+
+	err := uri.AddPathAttributeUnencoded("slot-id", []byte("12345"))
```

Comment:
Bit of a boring test. The encoding on "12345" (`0x31 0x32 0x33 0x34 0x45`) does nothing, which I presume is why the test succeeded even when you were double-encoding? 

It might be better to test with actual percent signs in the values, like `id=%25%30%31` (three bytes `0x25 0x30 0x31` or the literal string `%01`). And with non-ASCII characters that *need* encoding, such as `id=%00` which is `[]byte{0}`.

Reply:
I modified the test now to use `the pin` as PIN.

Reply:
Unless there are other concerns I will merge this patch.

Reply:
Anyone got any comments on this patch? Otherwise I'd merge it soon.